node/network/availability-recovery/src/lib.rs

```diff
@@ -0,0 +1,724 @@
+	requesting_chunks: FuturesUnordered<oneshot::Receiver<ChunkResponse>>,
+}
+
+impl Interaction {
+	async fn run(mut self) {
+		loop {
+			// If it's empty and requesting_chunks is empty,
+			// break and issue a FromInteraction::Concluded(RecoveryError::Unavailable)
+			if self.requesting_chunks.is_empty() && self.shuffling.is_empty() {
+				if self.to_state.send(FromInteraction::Concluded(
+					self.candidate_hash,
+					Err(RecoveryError::Unavailable),
+				)).await.is_err() {
+					return;
+				};
+				break;
+			}
+
+			while self.requesting_chunks.len() < N_PARALLEL {
+				if let Some(validator_index) = self.shuffling.pop() {
+					let (tx, rx) = oneshot::channel();
+
+					if self.to_state.send(FromInteraction::MakeRequest(
+						self.validators[validator_index as usize].clone(),
+						self.candidate_hash.clone(),
+						validator_index,
+						tx,
+					)).await.is_err() {
+						return;
+					};
+
+					self.requesting_chunks.push(rx);
+				} else {
+					break;
+				}
+			}
+
+			// Poll for new updates from requesting_chunks.
+			if let Some(Ok(Ok((peer_id, chunk)))) = self.requesting_chunks.next().await {
+				// Check merkle proofs of any received chunks, and any failures should
+				// lead to issuance of a FromInteraction::ReportPeer message.
+				if let Ok(anticipated_hash) = branch_hash(
+					&self.erasure_root,
+					&chunk.proof,
+					chunk.index as usize,
+				) {
+					let erasure_chunk_hash = BlakeTwo256::hash(&chunk.chunk);
+
+					if erasure_chunk_hash != anticipated_hash {
+						if self.to_state.send(FromInteraction::ReportPeer(
+							peer_id,
+							COST_MERKLE_PROOF_INVALID,
+						)).await.is_err() {
+							return;
+						}
+					} else {
+						self.received_chunks.insert(peer_id, chunk);
+					}
+				} else {
+					if self.to_state.send(FromInteraction::ReportPeer(
+						peer_id,
+						COST_MERKLE_PROOF_INVALID,
+					)).await.is_err() {
+						return;
+					}
+				}
+			}
+
+			// If received_chunks has more than threshold entries, attempt to recover the data.
+			// If that fails, or a re-encoding of it doesn't match the expected erasure root,
+			// break and issue a FromInteraction::Concluded(RecoveryError::Invalid).
+			// Otherwise, issue a FromInteraction::Concluded(Ok(())).
+			if self.received_chunks.len() >= self.threshold {
+				let concluded = match polkadot_erasure_coding::reconstruct_v1(
+					self.validators.len(),
+					self.received_chunks.values().map(|c| (&c.chunk[..], c.index as usize)),
+				) {
+					Ok(data) => FromInteraction::Concluded(self.candidate_hash.clone(), Ok(data)),
+					Err(_) => FromInteraction::Concluded(self.candidate_hash.clone(), Err(RecoveryError::Invalid)),
+				};
+
+				if let Err(e) = self.to_state.send(concluded).await {
+					tracing::warn!(
+						target: LOG_TARGET,
+						err = ?e,
+						"Failed to send Concluded result from interaction",
+					);
+				}
+				return;
+			}
+		}
+	}
+}
+
+struct State {
```

Comment:
This struct differs from the guide: the guide has `connecting_validators`, while this has `connection_requests`. These have different types, but I suspect that they serve the same purpose. Either the guide or this struct should be updated so they're in sync.

Reply:
The types are actually the same `ConnectionRequests` is our implementation of the stream mentioned there. But yeah, the names appear different.